File: ios/NaverMapView.swift
import UIKit
import NMapsMap
import React

private struct NaverPlaceMarker: Equatable {
    let id: String
    let category: String
    let lat: Double
    let lng: Double
}

@objc(NaverMapView)
final class NaverMapView: UIView, NMFMapViewCameraDelegate, NMFMapViewTouchDelegate {
    // pod install 시 .env의 NAVER_MAP_CLIENT_ID → NaverMap.generated.xcconfig를 생성한다.
    // Debug/Release 설정을 거쳐 Info.plist의 NMFNcpKeyId로 들어가며 SDK가 이 값으로 인증한다.
    // iOS Bundle ID는 네이버 클라우드 Maps 앱에 등록해야 한다. Client Secret은 사용하지 않는다.
    private let mapView = NMFMapView(frame: .zero)
    private var placeData: [NaverPlaceMarker] = []
    private var placeMarkers: [String: NMFMarker] = [:]
    private var icons: [String: NMFOverlayImage] = [:]
    private let userMarker = NMFMarker()
    private var markersDirty = true
    private var lastCamera: (lat: Double, lng: Double, zoom: Double)?
    private var lastFollowUser = false
    private var appliedNightMode: Bool?

    @objc var centerLat: NSNumber?
    @objc var centerLng: NSNumber?
    @objc var zoomLevel: NSNumber?
    @objc var userLat: NSNumber?
    @objc var userLng: NSNumber?
    @objc var followUser = true
    @objc var nightMode = false
    @objc var markers: NSArray? {
        didSet {
            let next = parseMarkers(markers)
            if next != placeData { placeData = next; markersDirty = true }
        }
    }
    @objc var onCameraIdle: RCTDirectEventBlock?
    @objc var onMarkerPress: RCTDirectEventBlock?

    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(mapView)
        mapView.addCameraDelegate(delegate: self)
        mapView.touchDelegate = self
        // 로고는 숨기거나 다른 UI로 가리지 않고, 검색/카테고리 아래로 위치와 여백만 조정한다.
        // https://navermaps.github.io/ios-map-sdk/guide-ko/4-1.html
        mapView.logoAlign = .leftTop
        mapView.logoMargin = UIEdgeInsets(top: 160, left: 16, bottom: 0, right: 0)
        userMarker.iconImage = NMFOverlayImage(image: makeUserLocationImage())
        userMarker.anchor = CGPoint(x: 0.5, y: 0.72)
        userMarker.zIndex = 20
        userMarker.isForceShowIcon = true
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    override func layoutSubviews() {
        super.layoutSubviews()
        mapView.frame = bounds
    }

    // .m에 공개한 props를 RN이 모두 설정한 뒤 마커와 카메라를 한 번에 갱신한다.
    // 위도만 바뀐 중간 좌표로 이동하는 것을 막으며, 이후 SDK 이벤트는 RCTDirectEventBlock으로 JS에 전달한다.
    override func didSetProps(_ changedProps: [String]!) {
        super.didSetProps(changedProps)
        applyProps()
    }

    func mapViewCameraIdle(_ mapView: NMFMapView) {
        let target = mapView.cameraPosition.target
        onCameraIdle?(["lat": target.lat, "lng": target.lng])
    }

    func mapView(_ mapView: NMFMapView, didTapMap latlng: NMGLatLng, point: CGPoint) {
        // Use screen distance so the touch target stays the same size at every zoom level.
        let nearest = placeData.map { marker -> (String, CGFloat) in
            let screen = mapView.projection.point(from: NMGLatLng(lat: marker.lat, lng: marker.lng))
            let dx = screen.x - point.x
            let dy = screen.y - point.y
            return (marker.id, dx * dx + dy * dy)
        }.filter { $0.1 <= 54 * 54 }.min { $0.1 < $1.1 }
        if let nearest { onMarkerPress?(["markerId": nearest.0]) }
    }

    private func applyProps() {
        if appliedNightMode != nightMode {
            // 네이버 SDK 야간 모드는 Navi 유형에서 지원한다. 라이트 모드에서는 Basic으로 복원한다.
            // 같은 지도 뷰에서 스타일만 변경하여 카메라와 마커 선택 상태를 유지한다.
            mapView.mapType = nightMode ? .navi : .basic
            mapView.isNightModeEnabled = nightMode
            appliedNightMode = nightMode
        }
        if markersDirty {
            let ids = Set(placeData.map { $0.id })
            for id in Array(placeMarkers.keys) where !ids.contains(id) {
                let marker = placeMarkers.removeValue(forKey: id)
                marker?.touchHandler = nil
                marker?.mapView = nil
            }
            for data in placeData {
                let marker = placeMarkers[data.id] ?? NMFMarker()
                marker.position = NMGLatLng(lat: data.lat, lng: data.lng)
                marker.iconImage = placeIcon(data.category)
                marker.anchor = CGPoint(x: 0.5, y: 0.62)
                marker.isHideCollidedMarkers = false
                marker.isHideCollidedSymbols = false
                marker.isForceShowIcon = true
                marker.zIndex = 10
                marker.touchHandler = { [weak self] _ in
                    self?.onMarkerPress?(["markerId": data.id])
                    return true
                }
                marker.mapView = mapView
                placeMarkers[data.id] = marker
            }
            markersDirty = false
        }
        let userLat = userLat?.doubleValue
        let userLng = userLng?.doubleValue
        let validUser = validCoordinate(userLat, userLng)
        if validUser, let lat = userLat, let lng = userLng {
            userMarker.position = NMGLatLng(lat: lat, lng: lng)
            userMarker.mapView = mapView
        } else {
            userMarker.mapView = nil
        }
        let lat = followUser && validUser ? userLat : centerLat?.doubleValue
        let lng = followUser && validUser ? userLng : centerLng?.doubleValue
        let zoom = min(21, max(0, zoomLevel?.doubleValue ?? 17))
        if validCoordinate(lat, lng), let lat, let lng {
            let changed = lastCamera.map { $0.lat != lat || $0.lng != lng || $0.zoom != zoom } ?? true
            if changed || (followUser && !lastFollowUser) {
                let update = NMFCameraUpdate(scrollTo: NMGLatLng(lat: lat, lng: lng), zoomTo: zoom)
                if lastCamera != nil { update.animation = .easeIn; update.animationDuration = 0.3 }
                mapView.moveCamera(update)
                lastCamera = (lat, lng, zoom)
            }
        }
        lastFollowUser = followUser
    }

    private func validCoordinate(_ lat: Double?, _ lng: Double?) -> Bool {
        guard let lat, let lng else { return false }
        return lat.isFinite && lng.isFinite && (-90...90).contains(lat) && (-180...180).contains(lng)
    }

    private func parseMarkers(_ value: NSArray?) -> [NaverPlaceMarker] {
        (value ?? []).enumerated().compactMap { index, raw in
            guard let item = raw as? NSDictionary,
                  let lat = item["lat"] as? Double, let lng = item["lng"] as? Double,
                  validCoordinate(lat, lng) else { return nil }
            let category = item["category"] as? String ?? "etc"
            let categories = ["art", "beauty", "cafe", "etc", "fashion", "food", "game", "heritage", "music", "popup"]
            return NaverPlaceMarker(id: item["id"] as? String ?? "marker-\(index)",
                                    category: categories.contains(category) ? category : "etc", lat: lat, lng: lng)
        }
    }

    private func placeIcon(_ category: String) -> NMFOverlayImage {
        if let cached = icons[category] { return cached }
        let image = UIImage(named: "map_marker_default_\(category)")
            ?? UIImage(named: "map_marker_default_etc")!
        let icon = NMFOverlayImage(image: image)
        icons[category] = icon
        return icon
    }

    deinit {
        mapView.removeCameraDelegate(delegate: self)
        placeMarkers.values.forEach { $0.touchHandler = nil; $0.mapView = nil }
        userMarker.mapView = nil
    }
    private func makeUserLocationImage() -> UIImage {
        let size = CGSize(width: 30, height: 40)
        let renderer = UIGraphicsImageRenderer(size: size)

        return renderer.image { context in
            let cgContext = context.cgContext
            let pink = UIColor(red: 1.0, green: 0.098, blue: 0.337, alpha: 1.0)
            let white = UIColor.white
            let centerX = size.width / 2
            let circleCenter = CGPoint(x: centerX, y: 27)

            let arrowPath = UIBezierPath()
            arrowPath.move(to: CGPoint(x: centerX, y: 3))
            arrowPath.addLine(to: CGPoint(x: centerX - 9.5, y: 16.5))
            arrowPath.addLine(to: CGPoint(x: centerX + 9.5, y: 16.5))
            arrowPath.close()

            white.setStroke()
            pink.setFill()
            arrowPath.lineWidth = 2.5
            arrowPath.lineJoinStyle = .round
            arrowPath.stroke()
            arrowPath.fill()

            cgContext.saveGState()
            cgContext.setShadow(offset: CGSize(width: 0, height: 2), blur: 6, color: UIColor.black.withAlphaComponent(0.16).cgColor)
            white.setFill()
            UIBezierPath(
                ovalIn: CGRect(
                    x: circleCenter.x - 13,
                    y: circleCenter.y - 13,
                    width: 26,
                    height: 26
                )
            ).fill()
            cgContext.restoreGState()

            pink.setFill()
            UIBezierPath(
                ovalIn: CGRect(
                    x: circleCenter.x - 10.5,
                    y: circleCenter.y - 10.5,
                    width: 21,
                    height: 21
                )
            ).fill()
        }
    }
}

File: android/app/src/main/java/com/rmdka/pingdomapp/NaverMapView.kt
package com.rmdka.pingdomapp

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PointF
import android.view.Gravity
import android.widget.FrameLayout
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.common.LifecycleState
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.UIManagerHelper
import com.facebook.react.uimanager.events.Event
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.CameraAnimation
import com.naver.maps.map.CameraUpdate
import com.naver.maps.map.MapView
import com.naver.maps.map.NaverMap
import com.naver.maps.map.NaverMapOptions
import com.naver.maps.map.overlay.Marker
import com.naver.maps.map.overlay.OverlayImage
import kotlin.math.roundToInt

private data class NaverPlaceMarker(val id: String, val category: String, val lat: Double, val lng: Double)

/** Native V2 host. Props are applied together to avoid moving to a half-updated coordinate. */
class NaverMapView(private val reactContext: ThemedReactContext) :
    FrameLayout(reactContext), LifecycleEventListener {
    // Keep SurfaceView rendering so MapGlassBackdropView can capture it with PixelCopy.
    private val mapView = MapView(reactContext, NaverMapOptions().useTextureView(false))
    private var naverMap: NaverMap? = null
    private var centerLat: Double? = null
    private var centerLng: Double? = null
    private var userLat: Double? = null
    private var userLng: Double? = null
    private var followUser = true
    private var nightMode = false
    private var appliedNightMode: Boolean? = null
    private var zoomLevel = 17
    private var placeData = emptyList<NaverPlaceMarker>()
    private var markersDirty = true
    private val placeMarkers = mutableMapOf<String, Marker>()
    private val icons = mutableMapOf<String, OverlayImage>()
    private val userMarker = Marker()
    private var lastCamera: Triple<Double, Double, Int>? = null
    private var lastFollowUser = false
    private var started = false
    private var resumed = false
    private var disposed = false
    private var hostResumed = reactContext.lifecycleState == LifecycleState.RESUMED

    // React Native owns the host layout; schedule a native child layout pass when
    // SDK controls (including attribution) request one after asynchronous map setup.
    private val layoutChildren = Runnable {
        if (!disposed && width > 0 && height > 0) {
            measure(MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY),
                    MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY))
            layout(left, top, right, bottom)
        }
    }

    init {
        addView(mapView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        mapView.onCreate(null)
        reactContext.addLifecycleEventListener(this)
        // SDK의 지도 제어 객체가 준비된 뒤에만 마커/카메라 API를 호출한다.
        // 준비 전에 들어온 React props는 필드에 보관했다가 마지막 applyProps()에서 반영한다.
        mapView.getMapAsync { map ->
            if (!disposed) {
                naverMap = map
                map.uiSettings.apply {
                    isZoomControlEnabled = false
                    isLocationButtonEnabled = false
                    isCompassEnabled = false
                    isScaleBarEnabled = false
                    // 공식 SDK 안내상 로고는 숨기거나 다른 UI로 가리지 않고 위치/여백만 조정한다.
                    // https://navermaps.github.io/android-map-sdk/guide-ko/4-1.html
                    // 검색창·카테고리 아래 왼쪽에 두어 우측의 앱 버튼과 겹치지 않게 한다.
                    logoGravity = Gravity.TOP or Gravity.LEFT
                    setLogoMargin(dp(16f), dp(160f), 0, 0)
                }
                map.addOnCameraIdleListener {
                    val target = map.cameraPosition.target
                    emit("topCameraIdle", Arguments.createMap().apply {
                        putDouble("lat", target.latitude)
                        putDouble("lng", target.longitude)
                    })
                }
                map.setOnMapClickListener { point, _ ->
                    // Maintain the existing forgiving touch area; direct marker clicks consume events.
                    val nearest = placeData.minByOrNull {
                        val screen = map.projection.toScreenLocation(LatLng(it.lat, it.lng))
                        val dx = screen.x - point.x
                        val dy = screen.y - point.y
                        dx * dx + dy * dy
                    }
                    nearest?.let {
                        val screen = map.projection.toScreenLocation(LatLng(it.lat, it.lng))
                        val dx = screen.x - point.x
                        val dy = screen.y - point.y
                        val tolerance = dp(54f).toFloat()
                        if (dx * dx + dy * dy <= tolerance * tolerance) emitMarkerPress(it.id)
                    }
                }
                applyProps()
            }
        }
    }

    fun setCenterLat(value: Double) { centerLat = value }
    fun setCenterLng(value: Double) { centerLng = value }
    fun setUserLat(value: Double?) { userLat = value }
    fun setUserLng(value: Double?) { userLng = value }
    fun setZoomLevel(value: Int) { zoomLevel = value.coerceIn(0, 21) }
    fun setFollowUser(value: Boolean) { followUser = value }
    fun setNightMode(value: Boolean) { nightMode = value }

    fun setMarkers(value: ReadableArray?) {
        val next = buildList {
            for (index in 0 until (value?.size() ?: 0)) {
                val item = value?.getMap(index) ?: continue
                if (!item.hasKey("lat") || item.isNull("lat") || !item.hasKey("lng") || item.isNull("lng")) continue
                val lat = item.getDouble("lat")
                val lng = item.getDouble("lng")
                if (!validCoordinate(lat, lng)) continue
                val id = if (item.hasKey("id")) item.getString("id") else null
                val category = if (item.hasKey("category")) item.getString("category") else null
                add(NaverPlaceMarker(id ?: "marker-$index", normalizeCategory(category), lat, lng))
            }
        }
        if (next != placeData) {
            placeData = next
            markersDirty = true
        }
    }

    // 앱의 마커 ID를 SDK Marker와 연결한다. SDK 터치 시 같은 ID를 JS에 돌려줘 장소를 선택한다.
    // 현재 위치도 앱에서 받은 좌표로 표시하며, SDK가 별도로 GPS 권한을 요청하지 않는다.
    fun applyProps() {
        if (disposed) return
        val map = naverMap ?: return
        if (appliedNightMode != nightMode) {
            // Basic 유형에는 야간 모드가 적용되지 않으므로 다크 테마에서 Navi로 전환한다.
            // 지도 뷰를 재생성하지 않아 카메라와 마커 선택 상태는 유지된다.
            map.mapType = if (nightMode) NaverMap.MapType.Navi else NaverMap.MapType.Basic
            map.isNightModeEnabled = nightMode
            appliedNightMode = nightMode
        }
        if (markersDirty) {
            val ids = placeData.map { it.id }.toSet()
            placeMarkers.keys.filter { it !in ids }.forEach { id ->
                placeMarkers.remove(id)?.apply { this.map = null; onClickListener = null }
            }
            for (data in placeData) {
                val marker = placeMarkers.getOrPut(data.id) { Marker() }
                marker.position = LatLng(data.lat, data.lng)
                marker.icon = placeIcon(data.category)
                marker.anchor = PointF(0.5f, 0.62f)
                marker.isHideCollidedMarkers = false
                marker.isHideCollidedSymbols = false
                marker.isForceShowIcon = true
                marker.zIndex = 10
                marker.setOnClickListener { emitMarkerPress(data.id); true }
                marker.map = map
            }
            markersDirty = false
        }
        val validUser = validCoordinate(userLat, userLng)
        if (validUser) {
            userMarker.position = LatLng(userLat!!, userLng!!)
            if (userMarker.map == null) {
                userMarker.icon = OverlayImage.fromBitmap(createUserLocationBitmap())
                userMarker.anchor = PointF(0.5f, userAnchorY())
                userMarker.zIndex = 20
                userMarker.isForceShowIcon = true
                userMarker.map = map
            }
        } else {
            userMarker.map = null
        }
        val lat = if (followUser && validUser) userLat else centerLat
        val lng = if (followUser && validUser) userLng else centerLng
        if (validCoordinate(lat, lng)) {
            val next = Triple(lat!!, lng!!, zoomLevel)
            if (next != lastCamera || (followUser && !lastFollowUser)) {
                val update = CameraUpdate.scrollAndZoomTo(LatLng(lat, lng), zoomLevel.toDouble())
                if (lastCamera != null) update.animate(CameraAnimation.Easing, 300)
                map.moveCamera(update)
                lastCamera = next
            }
        }
        lastFollowUser = followUser
    }

    private fun placeIcon(category: String): OverlayImage = icons.getOrPut(category) {
        val resourceId = resources.getIdentifier("map_marker_default_$category", "drawable", reactContext.packageName)
        val original = BitmapFactory.decodeResource(resources, if (resourceId != 0) resourceId else R.drawable.map_marker_default_etc)
        val width = dp(32f)
        val height = (original.height * width.toFloat() / original.width).roundToInt().coerceAtLeast(1)
        val scaled = Bitmap.createScaledBitmap(original, width, height, true)
        if (scaled !== original) original.recycle()
        OverlayImage.fromBitmap(scaled)
    }

    private fun normalizeCategory(value: String?): String = when (value) {
        "art", "beauty", "cafe", "etc", "fashion", "food", "game", "heritage", "music", "popup" -> value
        else -> "etc"
    }

    private fun validCoordinate(lat: Double?, lng: Double?): Boolean =
        lat != null && lng != null && lat.isFinite() && lng.isFinite() && lat in -90.0..90.0 && lng in -180.0..180.0

    private fun dp(value: Float) = (value * resources.displayMetrics.density).roundToInt()

    private fun emitMarkerPress(markerId: String) = emit("topMarkerPress", Arguments.createMap().apply {
        putString("markerId", markerId)
    })

    private fun emit(name: String, data: WritableMap) {
        if (disposed || id == NO_ID) return
        UIManagerHelper.getEventDispatcherForReactTag(reactContext, id)?.dispatchEvent(
            NaverMapEvent(name, UIManagerHelper.getSurfaceId(this), id, data)
        )
    }

    override fun requestLayout() {
        super.requestLayout()
        if (width <= 0 || height <= 0) return
        removeCallbacks(layoutChildren)
        post(layoutChildren)
    }

    override fun onAttachedToWindow() { super.onAttachedToWindow(); updateLifecycle() }
    override fun onDetachedFromWindow() { pauseMap(); super.onDetachedFromWindow() }
    override fun onHostResume() { hostResumed = true; updateLifecycle() }
    override fun onHostPause() { hostResumed = false; pauseMap() }
    override fun onHostDestroy() { dispose() }

    // RN 화면 부착 상태와 앱 foreground 상태를 SDK 수명주기에 전달한다.
    // 화면 이탈/백그라운드에서는 pauseMap(), 실제 뷰 폐기 시에는 dispose()로 렌더링 자원을 정리한다.
    private fun updateLifecycle() {
        if (disposed || !isAttachedToWindow || !hostResumed) return
        if (!started) { mapView.onStart(); started = true }
        if (!resumed) { mapView.onResume(); resumed = true }
    }

    private fun pauseMap() {
        if (resumed) { mapView.onPause(); resumed = false }
        if (started) { mapView.onStop(); started = false }
    }

    fun dispose() {
        if (disposed) return
        disposed = true
        removeCallbacks(layoutChildren)
        reactContext.removeLifecycleEventListener(this)
        placeMarkers.values.forEach { it.map = null; it.onClickListener = null }
        placeMarkers.clear()
        userMarker.map = null
        pauseMap()
        mapView.onDestroy()
        naverMap = null
        icons.clear()
    }

    private fun userAnchorY(): Float {
        val scale = resources.displayMetrics.density
        val height = (40 * scale).roundToInt()
        val top = (3 * scale).roundToInt()
        return (top + height * 0.72f) / (height + top + scale.roundToInt())
    }
    private fun createUserLocationBitmap(): Bitmap {
        val scale = resources.displayMetrics.density
        val width = (30 * scale).toInt()
        val height = (40 * scale).toInt()
        val centerX = width / 2f
        val circleCenterY = 27 * scale
        val circleRadius = 10.5f * scale

        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)

        val shadowPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.argb(45, 0, 0, 0)
            style = Paint.Style.FILL
        }
        val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = 0xFFFF1956.toInt()
            style = Paint.Style.FILL
        }
        val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.WHITE
            style = Paint.Style.STROKE
            strokeJoin = Paint.Join.ROUND
            strokeCap = Paint.Cap.ROUND
            strokeWidth = 2.5f * scale
        }

        val arrowPath = Path().apply {
            moveTo(centerX, 3f * scale)
            lineTo(centerX - 9.5f * scale, 16.5f * scale)
            lineTo(centerX + 9.5f * scale, 16.5f * scale)
            close()
        }

        canvas.drawCircle(centerX, circleCenterY + scale, circleRadius + 2.5f * scale, shadowPaint)
        canvas.drawPath(arrowPath, strokePaint)
        canvas.drawPath(arrowPath, fillPaint)
        canvas.drawCircle(centerX, circleCenterY, circleRadius + 2.5f * scale, Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.WHITE
            style = Paint.Style.FILL
        })
        canvas.drawCircle(centerX, circleCenterY, circleRadius, fillPaint)

        return withBitmapPadding(
            source = bitmap,
            topPaddingPx = (3f * scale).roundToInt(),
            bottomPaddingPx = scale.roundToInt(),
            sidePaddingPx = scale.roundToInt()
        )
    }

    private fun withBitmapPadding(
        source: Bitmap,
        topPaddingPx: Int = 0,
        bottomPaddingPx: Int = 0,
        sidePaddingPx: Int = 0
    ): Bitmap {
        if (topPaddingPx == 0 && bottomPaddingPx == 0 && sidePaddingPx == 0) {
            return source
        }

        val paddedBitmap = Bitmap.createBitmap(
            source.width + sidePaddingPx * 2,
            source.height + topPaddingPx + bottomPaddingPx,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(paddedBitmap)
        canvas.drawBitmap(source, sidePaddingPx.toFloat(), topPaddingPx.toFloat(), null)
        return paddedBitmap
    }

}

private class NaverMapEvent(
    private val name: String,
    surfaceId: Int,
    viewId: Int,
    private val data: WritableMap,
) : Event<NaverMapEvent>(surfaceId, viewId) {
    override fun getEventName() = name
    override fun canCoalesce() = false
    override fun getEventData() = data
}
